File: packages/core/src/core/math.ts
/** Small, dependency-free math helpers. */

export type RGB = [number, number, number];

export const clamp = (v: number, lo: number, hi: number): number =>
  v < lo ? lo : v > hi ? hi : v;

export const lerp = (a: number, b: number, t: number): number => a + (b - a) * t;

/** distance between two points. */
export const dist = (ax: number, ay: number, bx: number, by: number): number =>
  Math.hypot(ax - bx, ay - by);

/** parse `#rrggbb` (or `#rgb`) → `[r, g, b]`, falling back to a cool blue. */
export function hexToRgb(hex: string): RGB {
  const h = hex.replace('#', '');
  const x = h.length === 3 ? h.replace(/./g, (c) => c + c) : h;
  const n = Number.parseInt(x.slice(0, 6), 16);
  if (Number.isNaN(n)) return [77, 163, 255];
  return [(n >> 16) & 255, (n >> 8) & 255, n & 255];
}

/** The resting (cool) and energized (warm) ends of the free-particle ramp (§20.8). */
export const COOL: RGB = [200, 224, 255];
export const WARM: RGB = [255, 122, 69];

/**
 * Free-particle color: cool centre → warm edge by `rs` (= normalized dist²),
 * then blended toward the travelling `accent` by `heat` (§20.8 / the prototype).
 */
export function particleRGB(rs: number, heat: number, accent: RGB): RGB {
  return particleRGBInto([0, 0, 0], rs, heat, accent);
}

/** `particleRGB` writing into a caller-owned `out` — identical color, zero allocation. The draw
 *  loop calls this with a shared scratch so it doesn't allocate an `[r,g,b]` per particle per
 *  frame. `out` is returned for convenience. */
export function particleRGBInto(out: RGB, rs: number, heat: number, accent: RGB): RGB {
  let r = COOL[0] + (WARM[0] - COOL[0]) * rs;
  let g = COOL[1] + (WARM[1] - COOL[1]) * rs;
  let b = COOL[2] + (WARM[2] - COOL[2]) * rs;
  r += (accent[0] - r) * heat;
  g += (accent[1] - g) * heat;
  b += (accent[2] - b) * heat;
  out[0] = r;
  out[1] = g;
  out[2] = b;
  return out;
}

/** `[r, g, b]` → `#rrggbb`. */
export function rgbToHex([r, g, b]: RGB): string {
  const h = (v: number): string => clamp(Math.round(v), 0, 255).toString(16).padStart(2, '0');
  return `#${h(r)}${h(g)}${h(b)}`;
}

/** Lerp two hex colors by `t` ∈ [0,1] — conserved pigment mixing (§20.8). */
export function mixHex(a: string, b: string, t: number): string {
  const [ar, ag, ab] = hexToRgb(a);
  const [br, bg, bb] = hexToRgb(b);
  const k = t < 0 ? 0 : t > 1 ? 1 : t;
  return rgbToHex([ar + (br - ar) * k, ag + (bg - ag) * k, ab + (bb - ab) * k]);
}

/** Sample a color ramp at `frac` ∈ [0,1] — the accent journey (§9). */
export function sampleStops(stops: readonly RGB[], frac: number): RGB {
  if (stops.length === 0) return [77, 163, 255];
  if (stops.length === 1) return stops[0]!;
  const f = clamp(frac, 0, 1) * (stops.length - 1);
  const i = Math.min(stops.length - 2, Math.floor(f));
  const t = f - i;
  const a = stops[i]!;
  const b = stops[i + 1]!;
  return [a[0] + (b[0] - a[0]) * t, a[1] + (b[1] - a[1]) * t, a[2] + (b[2] - a[2]) * t];
}

/**
 * The `screen` attenuation factor (workover v0.3 §"`screen` modifier") — pure:
 *
 *   falloff      = max(0, 1 − d/range)²
 *   screenFactor = clamp(1 − S·falloff, min, 1)
 *
 * A body carrying the `screen` token damps OTHER bodies' forces on matter within its
 * range by this factor (applied in the integrator's force pass). Smooth at the edge
 * (falloff → 0 as d → range; no hard cliff); `min` clamps the floor (`data-screen-min`,
 * default 0 = full cancellation at the core is allowed); a non-positive range is inert
 * (returns 1 — screens are always local, never global), which also guarantees no NaN
 * at zero range.
 */
export function screenFactor(d: number, range: number, strength: number, min = 0): number {
  if (!(range > 0)) return 1;
  const fall = Math.max(0, 1 - d / range);
  const factor = 1 - strength * fall * fall;
  const floor = Math.min(Math.max(min, 0), 1);
  return factor < floor ? floor : factor > 1 ? 1 : factor;
}

File: swift/Sources/FieldUICore/Math/Math.swift
import Foundation
import simd

// MARK: - Primitive aliases

/// The physics vector throughout the engine. 3D-native: on 2D platforms z=0.
public typealias Vec3 = SIMD3<Float>

/// An RGB triple ∈ [0,255].
public typealias RGB = SIMD3<Float>

// MARK: - Scalar helpers

@inline(__always)
public func clamp(_ v: Float, _ lo: Float, _ hi: Float) -> Float {
    v < lo ? lo : v > hi ? hi : v
}

@inline(__always)
public func lerp(_ a: Float, _ b: Float, _ t: Float) -> Float {
    a + (b - a) * t
}

// MARK: - Color

/// Cool (resting) particle color (§20.8).
public let COOL: RGB = RGB(200, 224, 255)
/// Warm (energized) particle color (§20.8).
public let WARM: RGB = RGB(255, 122, 69)
/// Fallback accent blue.
public let DEFAULT_ACCENT: RGB = RGB(77, 163, 255)

/// Parse `#rrggbb` or `#rgb` → RGB, falling back to the default blue.
public func hexToRgb(_ hex: String) -> RGB {
    var h = hex.hasPrefix("#") ? String(hex.dropFirst()) : hex
    if h.count == 3 {
        h = h.flatMap { [String($0), String($0)] }.joined()
    }
    guard h.count >= 6, let n = UInt32(h.prefix(6), radix: 16) else {
        return DEFAULT_ACCENT
    }
    return RGB(
        Float((n >> 16) & 0xFF),
        Float((n >>  8) & 0xFF),
        Float( n        & 0xFF)
    )
}

/// RGB → `#rrggbb`.
public func rgbToHex(_ c: RGB) -> String {
    func h(_ v: Float) -> String {
        String(format: "%02x", Int(clamp(v.rounded(), 0, 255)))
    }
    return "#\(h(c.x))\(h(c.y))\(h(c.z))"
}

/// Free-particle color: cool→warm by `rs` (normalized dist²), blended toward `accent` by `heat`.
public func particleRGB(rs: Float, heat: Float, accent: RGB) -> RGB {
    var c = COOL + (WARM - COOL) * rs
    c += (accent - c) * heat
    return c
}

/// Lerp two hex colors by `t` ∈ [0,1].
public func mixHex(_ a: String, _ b: String, t: Float) -> String {
    let ca = hexToRgb(a)
    let cb = hexToRgb(b)
    let k = clamp(t, 0, 1)
    return rgbToHex(ca + (cb - ca) * k)
}

/// Sample a color ramp at `frac` ∈ [0,1].
public func sampleStops(_ stops: [RGB], frac: Float) -> RGB {
    guard !stops.isEmpty else { return DEFAULT_ACCENT }
    if stops.count == 1 { return stops[0] }
    let f = clamp(frac, 0, 1) * Float(stops.count - 1)
    let i = min(stops.count - 2, Int(f))
    let t = f - Float(i)
    return stops[i] + (stops[i + 1] - stops[i]) * t
}

/// Screen attenuation factor (workover v0.3 §"`screen` modifier").
public func screenFactor(d: Float, range: Float, strength: Float, min floor: Float = 0) -> Float {
    guard range > 0 else { return 1 }
    let fall = max(0, 1 - d / range)
    let factor = 1 - strength * fall * fall
    let f = clamp(floor, 0, 1)
    return clamp(factor, f, 1)
}
